### PR TITLE
[FIXED JENKINS-40462] Move wrappers into options

### PR DESCRIPTION
**Pull request by abayer**
[JENKINS-40462](https://issues.jenkins-ci.org/browse/JENKINS-40462)

This depends on
https://github.com/jenkinsci/pipeline-model-definition-plugin/pull/70
so shouldn't be merged until that's in.

I'm not a huge fan of the hoops we have to jump through to support
wrappers in the same section as Describables, so I may tweak this.

cc @reviewbybees esp @rsandell

### CODE REVIEW

**Comment by rsandell** [comment]
I'm holding off my review until #70 is merged.

**Comment by abayer** [comment]
@rsandell Mind reviewing?

File: SYNTAX.md [hunk 1]
```diff
@@ -269,8 +269,8 @@ parameters {
 ```
 
 ### Options
-* *Description*: Other options exclusive to Declarative, such as `skipDefaultCheckout`, and traditional `JobProperty`s, such as 
-build discarding, limiting concurrent builds, and more.
+* *Description*: Other options exclusive to Declarative, such as `skipDefaultCheckout`, as well as traditional `JobProperty`s, such as 
+build discarding, limiting concurrent builds, and more, and "wrappers" that should wrap the entire build, such as `timeout`.
```
Comment (rsandell) [inline SYNTAX.md]: I think I got an aneurysm reading this sentence.

File: pipeline-model-definition/src/main/groovy/org/jenkinsci/plugins/pipeline/modeldefinition/model/Options.groovy [hunk 2]
```diff
@@ -72,31 +83,52 @@ public class Options implements Serializable {
...
 
+    public static Map<String,String> getEligibleWrapperStepClasses() {
+        Map<String,String> stepNames = [:]
+
+        // TODO: Figure out if we want to support metasteps? Don't think so.
+        ExtensionList.lookup(StepDescriptor.class).each { s ->
+            if (s.takesImplicitBlockArgument()
+                && !(s.getFunctionName() in ModelASTMethodCall.blockedSteps.keySet())
+                && !(Launcher.class in s.getRequiredContext())
+                && !(FilePath.class in s.getRequiredContext())) {
```
Comment (rsandell) [inline pipeline-model-definition/src/main/groovy/org/jenkinsci/plugins/pipeline/modeldefinition/model/Options.groovy]: I'm not sure about the cost of calling `s.getRequiredContext()` but if it does some kind of lookup you should consider cashing it in a local variable first.

File: pipeline-model-definition/src/main/groovy/org/jenkinsci/plugins/pipeline/modeldefinition/parser/JSONParser.groovy [hunk 3]
```diff
@@ -113,9 +113,6 @@ class JSONParser implements Parser {
                 case 'parameters':
                     pipelineDef.parameters = parseBuildParameters(pipelineJson.getJSONObject("parameters"))
                     break
-                case 'wrappers':
```
Comment (rsandell) [inline pipeline-model-definition/src/main/groovy/org/jenkinsci/plugins/pipeline/modeldefinition/parser/JSONParser.groovy]: Deprecation warning?

File: SYNTAX.md (hunk 1 above)
Comment (abayer) [inline SYNTAX.md]: Rephrased. =)

File: pipeline-model-definition/src/main/groovy/org/jenkinsci/plugins/pipeline/modeldefinition/parser/JSONParser.groovy (hunk 3 above)
Comment (abayer) [inline pipeline-model-definition/src/main/groovy/org/jenkinsci/plugins/pipeline/modeldefinition/parser/JSONParser.groovy]: Nah - we don't really care about that for the JSON at this point. Plus, `wrappers` would actually fail at schema validation time, before we even got this far.

**Comment by abayer** [comment]
ping @rsandell

**Comment by abayer** [comment]
Gonna merge this - if it turns out there are problems with the new caching approach, we can fix them later.

**Pull request merged**